### PR TITLE
Fix support for Perl 5.6.2

### PR DESCRIPTION
**Pull request by pali**
These patches try to fix support for Perl 5.6.2.

Lot of modules are compatible with Perl 5.6.2, but their test suite uses Test::More. This could allow to use Test::More also for writing tests in modules which are still compatible with Perl 5.6.2.

### CODE REVIEW

File: lib/Test2/API.pm [hunk 1]
```diff
@@ -406,7 +406,7 @@ sub _existing_error {
     my $oldframe = $ctx->{trace}->frame;
     my $olddepth = $ctx->{_depth};
 
-    my $mess = longmess();
+    my $mess = Carp::longmess();
```
Comment (autarch) [inline lib/Test2/API.pm]: This sort of change really needs a comment or else someone is likely to revert it back later.

File: t/00-report.t [hunk 2]
```diff
@@ -67,7 +67,7 @@ my $len = 0;
 for my $dep (@depends) {
     my $l = length($dep);
     $len = $l if $l > $len;
-    $deps{$dep} = eval "require $dep; $dep->VERSION" || "N/A";
+    $deps{$dep} = eval "require $dep" ? ($dep->VERSION || '0') : 'N/A';
```
Comment (autarch) [inline t/00-report.t]: Why does this need to be changed for 5.6?

File: t/Legacy/Regression/736_use_ok.t [hunk 3]
```diff
@@ -18,13 +18,16 @@ sub capture(&) {
     return $warn || "";
 }
 
+{
+local $TODO = "known to fail on $]" if $] le "5.006002";
 my $file = __FILE__;
 my $line = __LINE__ + 2;
 like(
-    capture { use_ok 'MyWarner' },
+    capture { local $TODO; use_ok 'MyWarner' },
```
Comment (autarch) [inline t/Legacy/Regression/736_use_ok.t]: This could also use a comment. Why is this necessary?

File: t/Legacy_And_Test2/preload_diag_note.t [hunk 4]
```diff
@@ -1,6 +1,11 @@
 use strict;
 use warnings;
 
+if ($] lt "5.008000") {
```
Comment (autarch) [inline t/Legacy_And_Test2/preload_diag_note.t]: I don't think you need string comparison. You can just write `$] < 5.008`

File: t/00-report.t (hunk 2 above)
Comment (autarch) [reply t/00-report.t]: Ignore this comment. I get it.

File: lib/Test2/API.pm (hunk 1 above)
Comment (pali) [reply lib/Test2/API.pm]: I added comment also into code via git fixup commit.

File: t/00-report.t (hunk 2 above)
Comment (pali) [reply t/00-report.t]: No, it is not required for 5.6. But some modules (e.g. Carp) do not set `our $VERSION` and therefore that `$dep->VERSION` call returns empty string which is in boolean context treated as false and so `"N/A"` was used. This change is just for diagnostic purpose for any perl version to distinguish between fact that module was not loaded and fact that version of module was not specified.

File: t/Legacy/Regression/736_use_ok.t (hunk 3 above)
Comment (pali) [reply t/Legacy/Regression/736_use_ok.t]: I thought it is easy to understand reason... But I added explaining git fixup commit. It is better now?

File: t/Legacy_And_Test2/preload_diag_note.t (hunk 4 above)
Comment (pali) [reply t/Legacy_And_Test2/preload_diag_note.t]: It is a good idea to compare perl version via string comparison. Floating-point comparison depends on precision which can be in some cases lost. I was told that due to this fact for `$]` should be always used string version.

**Comment by pali** [comment]
> If I'm reading the logs right it tried to install dzil and its deps on 5.6.2 in order to build the distro.

This is problem in that custom perl travis script from repo https://github.com/travis-perl/helpers. I hacked travis to use Makefile.PL which is working fine.

> I can't imagine that's going to work.

So look, now tests passed!

File: t/Legacy_And_Test2/preload_diag_note.t (hunk 4 above)
Comment (pali) [reply t/Legacy_And_Test2/preload_diag_note.t]: Now I'm looking it is also written in perl documentation:
https://perldoc.perl.org/perlvar.html#%24%5d
> When comparing `$]` , string comparison operators are **highly recommended**. The inherent limitations of binary floating point representation can sometimes lead to incorrect comparisons for some numbers on some architectures.

Comment (haarg) [reply t/Legacy_And_Test2/preload_diag_note.t]: My recommendation would be to use `"$]" < 5.008000`.  It should really be a numeric comparison, but there are some issues with how `$]` is defined on earlier perl versions which forcing it through stringification will resolve.

**Comment by haarg** [comment]
In the failing travis test it isn't trying to install using perl 5.6.2, however it is trying to use a 5.6.2 compatible mirror.  The travis-perl helpers are generally meant to deal with this, but the handling of perl 5.6 is a bit rough so the attempt to use an earlier mirror doesn't work quite right in this case.

**Comment by pali** [comment]
Now I squashed changes and did some other fixes. Tests are passing.

**Comment by Leont** [comment]
> Honestly, I don't think 5.6.x support is really worthwhile. Even crufty distros and systems generally have 5.8 available.

Much of the 5.6 support around the toolchain is on a "patches welcome" base. Evidently it's still worthwhile for some people. Given Test::More's position in the river this doesn't surprise me.

**Comment by exodist** [comment]
I am not opposed to this, but I will not put any effort into maintaining 5.6 support. I just released the largest change to Test-Simple since the Test2 conversion. I want to let it brew on cpan for a while before I merge/release this.

**Comment by pali** [comment]
Ok, I could help if needed (in future).

Anyway, 2 weeks passed... do you want to wait for a longer time? Or is pull request ok?
